src/potato_cluster.erl

```diff
@@ -0,0 +1,233 @@
+-module(potato_cluster).
+
+-export([
+	 start_one_pop_verifier/5,
+	 start_single_server_cluster/5,
+	 start_cluster_from_json/2,
+	 stop_cluster/1,
+	 start_web_cluster/1
+	]).
+
+-include_lib("stdlib/include/assert.hrl").
+-include_lib("eunit/include/eunit.hrl").
+
+-include("potato_records.hrl").
+
+json_find(Key, Map) when is_atom(Key) ->
+    maps:find(atom_to_binary(Key, utf8), Map).
+
+json_get(Key, Map) when is_atom(Key) ->
+    maps:get(atom_to_binary(Key, utf8), Map).
+
+json_get_str(Key, Map) when is_atom(Key) ->
+    binary_to_list(maps:get(atom_to_binary(Key, utf8), Map)).
+
+json_put(Key, Value, Map) when is_atom(Key) ->
+    maps:put(atom_to_binary(Key, utf8), Value, Map).
+
+start_one_pop_verifier(VerifierArr, JsonConf, MyIndex, UdpServerId, OnEventFn) ->
+
+    MyAddress = (array:get(MyIndex, VerifierArr))#verifier_public_info.network_data,
+
+    {_NetworkAddress, MyNodeNetId} = MyAddress,
+
+    NetSendFn = fun(DestAddressList, MsgId, Data) -> 
+			gen_server:cast({global, UdpServerId}, {send, DestAddressList, {MyAddress, MsgId, Data} })
+		end,
+
+    ConfPrivateKey = default,
+
+    VerifierOnEventFn = fun(Code, Data) -> OnEventFn(MyNodeNetId, Code, Data) end,
+
+    ConfigData = {MyIndex, NetSendFn, VerifierOnEventFn, ConfPrivateKey},
+
+    gen_server:start_link({global, MyNodeNetId}, pop_verifier, {json_map, JsonConf, ConfigData}, []),
+
+    gen_server:cast({global, UdpServerId}, {add_node, MyNodeNetId, {global, MyNodeNetId}}),
+
+    MyNodeNetId.
+
+start_single_server_cluster(VerifierArr, JsonConf, ServerAddress, UdpServerId, OnEventFn) ->
+    {_Ip, Port} = ServerAddress,
+
+    UdpOnEventFn = fun(Code, Data) -> OnEventFn(UdpServerId, Code, Data) end,
+
+    gen_server:start_link({global, UdpServerId}, potato_udp, {Port, UdpOnEventFn}, []),
+
+    VerList = lists:filter(
+		fun(VerData) ->
+			{VerAddress, _} = VerData#verifier_public_info.network_data,
+			VerAddress == ServerAddress
+		end,
+		array:to_list(VerifierArr)),
+
+    IndexList = lists:map(
+		  fun(VerData) ->
+			  VerData#verifier_public_info.index
+		  end,
+		  VerList),
+
+    IdList = lists:map(
+		fun(VerIndex) ->
+			start_one_pop_verifier(VerifierArr, JsonConf, VerIndex, UdpServerId, OnEventFn)
+		end, 
+		IndexList),
+    IdList.
+
+
+start_cluster_from_json(JsonConf_0, OnEventFn) ->
+
+    JsonConf = 
+	case json_get(genesis_block_timestamp_sec, JsonConf_0) of
+
+	    <<"now">> ->
+		Time = erlang:system_time(second),
+		json_put(genesis_block_timestamp_sec, Time, JsonConf_0);
+
+	    _ ->
+		JsonConf_0
+	end,
+
+    
+    
+    VerifierArr = pop_verifier:make_verifier_array_from_json(JsonConf),
+
+    VerAddressList = lists:map(
+		       fun(VerData) ->
+			       {VerAddress, _} = VerData#verifier_public_info.network_data,
+			       VerAddress
+		       end,
+		       array:to_list(VerifierArr)),
+
+    VerAddressUniqueList = sets:to_list(sets:from_list(VerAddressList)),
+
+    ProcessData = lists:map(
```

Comment:
So this will start several clusters :o?  Is each cluster building its own blockchain?

Reply:
more like a cluster hosts multiple nodes behind the same udp server (multiple nodes for the same chain). in practice we'll probably have one node per server

Reply:
Oh, this is cool but why?

Reply:
so all the clusters host nodes that are all still verifying the same blockchain right now?

Reply:
yeah, just for testing so I don't start up a multiple udp servers

Reply:
🥔🥔🥔🥔 always better in clusters